### ISSUE TITLE
False E0605/E0308 on unsizing coercion when a supertrait has a negative const generic argument

### ISSUE THREAD

**Issue by vaivaswatha** [rust-lang/rust-analyzer#23373]
## Summary

When a trait has a supertrait whose const generic argument is a negative literal, rust-analyzer fails to prove that supertrait bound. Any unsizing coercion to that trait object is then rejected: `as &dyn Trait` reports a false `error[E0605]: non-primitive cast`, and `Box::new(value) in a Box<dyn Trait>` position reports a false `error[E0308]`. 

`cargo check` accepts all of the code.

An associated-const path (`{ i32::MAX }`) as the const argument fails the same way. A plain non-negative literal, or a free const item, works.
 
- **rust-analyzer version**: 0.3.3049-standalone
- **rustc version**: rustc 1.98.0 (88d9e12ae 2026-08-18), cargo 1.98.0, edition 2024
- **editor or extension**:  VS Code, rust-lang.rust-analyzer
- **relevant settings**: none (default configuration)
- **platform**: x86_64-unknown-linux-gnu

**code snippet to reproduce**:

```shell
cargo new --lib ra_issue
cd ra_issue
# paste the code below into src/lib.rs
cargo check                  # succeeds, no warnings
rust-analyzer diagnostics .  # reports two false errors
```

```rust
pub trait Base<const N: i32> {}

pub trait Neg: Base<-1> {}
pub trait Pos: Base<1> {}

#[derive(Clone, Copy)]
pub struct S;
impl Base<-1> for S {}
impl Base<1> for S {}
impl Neg for S {}
impl Pos for S {}

// BUG: error[E0605]: non-primitive cast: `&S` as `&(dyn Neg + 'static)`
pub fn cast_neg(s: &S) -> &dyn Neg {
    s as &dyn Neg
}

// No diagnostic. Identical except for the sign of the supertrait's const argument.
pub fn cast_pos(s: &S) -> &dyn Pos {
    s as &dyn Pos
}

// No diagnostic. Casting to the trait that carries the negative const argument
// directly is fine; only the supertrait obligation fails.
pub fn cast_base(s: &S) -> &dyn Base<-1> {
    s as &dyn Base<-1>
}

// BUG: error[E0605]. An associated-const path as the const argument fails too.
pub trait Assoc: Base<{ i32::MAX }> {}
impl Base<{ i32::MAX }> for S {}
impl Assoc for S {}
pub fn cast_assoc(s: &S) -> &dyn Assoc {
    s as &dyn Assoc
}

// No diagnostic. A free `const` item as the const argument is fine.
pub const END: i32 = 42;
pub trait Free: Base<END> {}
impl Base<END> for S {}
impl Free for S {}
pub fn cast_free(s: &S) -> &dyn Free {
    s as &dyn Free
}

// No diagnostic. An ordinary generic bound `T: Neg` resolves correctly,
// so only the unsizing coercion is affected.
pub fn takes_neg<T: Neg>(_t: &T) {}
pub fn call_takes_neg(s: &S) {
    takes_neg(s);
}

// BUG: error[E0308]: expected `Box<dyn Neg + 'static, Global>`, found `Box<S, Global>`
pub fn boxed_neg(s: &'static S) -> Box<dyn Neg> {
    Box::new(*s)
}
```

**Comment by ChayimFriedman2** [comment]
Fixed by https://github.com/rust-lang/rust-analyzer/pull/23179.
